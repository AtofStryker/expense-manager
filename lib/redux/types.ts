import { State } from '../state'
import { Action as ReduxAction } from 'redux'
import { ThunkAction } from 'redux-thunk'
import { Logger } from '../types'
import { Api } from '../api'

export interface ThunkExtraArgument {
  logger: Logger
  api: Api
}

export interface Action<Payload = void> extends ReduxAction<string> {
  loggable?: boolean
  payload?: Payload
  reducer: (state: State) => State
}

export type Thunk<Result = Promise<unknown>> = ThunkAction<
  Result,
  State,
  ThunkExtraArgument,
  Action<unknown>
>
