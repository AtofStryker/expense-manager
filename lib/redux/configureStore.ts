import { applyMiddleware, createStore } from 'redux'
import { createLogger } from 'redux-logger'
import thunk from 'redux-thunk'

import { Api } from '../api'
import { getInitialState } from '../state'
import { Logger } from '../types'

import rootReducer from './rootReducer'
import { Action, ThunkExtraArgument } from './types'

export const configureStore = () => {
  const logger: Logger = {
    log: (_, __) => null,
  }
  if (process.env.NODE_ENV === 'development') {
    logger.log = (message, payload) =>
      store.dispatch({
        type: message,
        payload,
      } as Action<any>)
  }

  const loggerMiddleware = createLogger({
    collapsed: true,
    predicate: (_, action: Action) => !(action.loggable === false),
    actionTransformer: (action: Action) => ({
      ...action,
      type: `${action.type}`,
    }),
  })

  const thunkExtra: ThunkExtraArgument = {
    logger,
    api: new Api(logger),
  }
  const middlewares = [thunk.withExtraArgument(thunkExtra)]
  if (process.env.NODE_ENV === 'development') {
    middlewares.push(loggerMiddleware)
  }

  const store = createStore(
    rootReducer as any,
    getInitialState() as any,
    applyMiddleware(...middlewares),
  )

  return store
}
