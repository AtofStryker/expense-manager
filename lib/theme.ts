// https://github.com/mui-org/material-ui/issues/13394
import { unstable_createMuiStrictModeTheme as createMuiTheme } from '@material-ui/core/styles'

// Create a theme instance.
const theme = createMuiTheme({
  palette: {
    primary: { main: '#a5790a' },
  },
})

export default theme
