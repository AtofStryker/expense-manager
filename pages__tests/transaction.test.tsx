import DateFnsUtils from '@date-io/date-fns'
import { MuiPickersUtilsProvider } from '@material-ui/pickers'
import { ThemeProvider } from '@material-ui/styles'
import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import renderer from 'react-test-renderer'
import { Store } from 'redux'

import { configureStore } from '../lib/redux/configureStore'
import theme from '../lib/theme'
import Transactions from '../pages/transactions'

const wrapWithProviders = (Component, store) => {
  return (
    <ReduxProvider store={store}>
      <ThemeProvider theme={theme}>
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <Component />
        </MuiPickersUtilsProvider>
      </ThemeProvider>
    </ReduxProvider>
  )
}

describe('With Snapshot Testing', () => {
  let store: Store

  beforeEach(() => {
    store = configureStore()
  })

  it('renders transactions', () => {
    const component = renderer.create(wrapWithProviders(Transactions, store))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
