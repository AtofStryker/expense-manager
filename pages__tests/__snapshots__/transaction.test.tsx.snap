// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Snapshot Testing renders transactions 1`] = `
Array [
  <div
    className="MuiBottomNavigation-root"
    style={
      Object {
        "bottom": 0,
        "position": "fixed",
        "width": "100%",
      }
    }
  >
    <button
      className="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "textTransform": "capitalize",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
        <span
          className="MuiBottomNavigationAction-label Mui-selected"
        >
          add
        </span>
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiBottomNavigationAction-root"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "textTransform": "capitalize",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"
          />
        </svg>
        <span
          className="MuiBottomNavigationAction-label"
        >
          transactions
        </span>
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiBottomNavigationAction-root"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "textTransform": "capitalize",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiBottomNavigationAction-wrapper"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h20v20H0V0z"
            fill="none"
            transform="scale(1.2, 1.2)"
          />
          <path
            d="M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
            transform="scale(1.2, 1.2)"
          />
        </svg>
        <span
          className="MuiBottomNavigationAction-label"
        >
          settings
        </span>
      </span>
    </button>
  </div>,
  "transactions",
]
`;
